From 3f3d6557c4bb09ea824518c5ddec8f82e915f1fc Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:46 -0400
Subject: [virt] kvm/nvmx: Fix up VM_ENTRY_IA32E_MODE control feature reporting

Message-id: <1406040016-3289-72-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86150
O-Subject: [RHEL7 PATCH v2 071/221] KVM: nVMX: Fix up VM_ENTRY_IA32E_MODE control feature reporting
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

Do not report that we can enter the guest in 64-bit mode if the host is
32-bit only. This is not supported by KVM.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 574353496081bce251fdd82fbfb67ca9fd6e0fcf)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 287fdc5..f2508d9 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -2212,9 +2212,13 @@ static __init void nested_vmx_setup_ctls_msrs(void)
  /* If bit 55 of VMX_BASIC is off, bits 0-8 and 12 must be 1. */
  nested_vmx_entry_ctls_low = VM_ENTRY_ALWAYSON_WITHOUT_TRUE_MSR;
  nested_vmx_entry_ctls_high &=
-  VM_ENTRY_LOAD_IA32_PAT | VM_ENTRY_IA32E_MODE;
+#ifdef CONFIG_X86_64
+  VM_ENTRY_IA32E_MODE |
+#endif
+  VM_ENTRY_LOAD_IA32_PAT;
  nested_vmx_entry_ctls_high |= (VM_ENTRY_ALWAYSON_WITHOUT_TRUE_MSR |
            VM_ENTRY_LOAD_IA32_EFER);
+
  /* cpu-based controls */
  rdmsr(MSR_IA32_VMX_PROCBASED_CTLS,
   nested_vmx_procbased_ctls_low, nested_vmx_procbased_ctls_high);
-- 
1.7.1